relationalai-sdk-snowflake/relationalai/snowflake_sdk/api.py

```diff
@@ -0,0 +1,269 @@
+            res:state::string         as state
+        from
+            (select GET_RAI_DATABASE('{database}') as res);
+    """)
+
+
+def list_databases(session: Session) -> DataFrame:
+    return session.sql("""
+        select
+            value:id::string           as id,
+            value:account_name::string as account_name,
+            value:created_by::string   as created_by,
+            value:name::string         as name,
+            value:region::string       as region,
+            value:state::string        as state
+        from
+            (select LIST_RAI_DATABASES() as res),
+            lateral flatten (input => res)
+    """)
+
+
+def use_database(session: Session, database: str) -> List[Row]:
+    res = session.sql(f"call USE_RAI_DATABASE('{database}')").collect()
+
+    if res[0][0] != database:
+        rsp = json.loads(res[0][0])
+
+        if not rsp["success"]:
+            raise Exception(rsp["message"])
+
+    return res
+
+
+def get_current_database(session: Session) -> DataFrame:
+    return session.sql("select CURRENT_RAI_DATABASE() as current_database")
+
+
+#################################
+# ENGINE
+#################################
+
+def create_engine(session: Session, engine: str, size: str = 'XS') -> List[Row]:
+    return session.sql(f"select CREATE_RAI_ENGINE('{engine}', '{size}') as status").collect()
+
+
+def delete_engine(session: Session, engine: str) -> List[Row]:
+    return session.sql(f"select DELETE_RAI_ENGINE('{engine}') as status").collect()
+
+
+def get_engine(session: Session, engine: str) -> DataFrame:
+    return session.sql(f"""
+        select
+            res:id::string           as id,
+            res:account_name::string as account_name,
+            res:created_by::string   as created_by,
+            res:created_on::string   as created_on,
+            res:name::string         as name,
+            res:region::string       as region,
+            res:size::string         as size,
+            res:state::string        as state
+        from
+            (select GET_RAI_ENGINE('{engine}') as res);
+    """)
+
+
+def list_engines(session: Session) -> DataFrame:
+    return session.sql(f"""
+        select
+            value:id::string           as id,
+            value:account_name::string as account_name,
+            value:created_by::string   as created_by,
+            value:created_on::string   as created_on,
+            value:name::string         as name,
+            value:region::string       as region,
+            value:size::string         as size,
+            value:state::string        as state
+        from
+            (select LIST_RAI_ENGINES() as res),
+            lateral flatten (input => res);
+""")
+
+
+def use_engine(session: Session, engine: str) -> List[Row]:
+    res = session.sql(f"CALL USE_RAI_ENGINE('{engine}')").collect()
+
+    if res[0][0] != engine:
+        rsp = json.loads(res[0][0])
+
+        if not rsp["success"]:
```

Comment:
This is the exact inconsistency we were talking about. The methods such as create_engine or create_database will fail with an error (aka exit on error) and their behavior doesn't depend on a variable but this one's does.

If we do some special handling we will need to do it for functions like get_engine because it looks like its behavior depends on the variable as well.

It looks to me like we need to first make the SQL library consistent and then we can use the variable for proper exception handling. Although, if we use a variable to control the behavior it makes it a bit weird e.g. for anyone working with SQL library directly (not including tests use case) they would want the variable to be set to true, but for anyone using the Python SDK they would want it to be set to false to allow Python SDK do the exceptions handling.

Reply:
Also, do you know if Snowflake has this kind of functionality. I don't expect it to have it though. Maybe we shouldn't have the variable either and do what we discussed previously by having a public interface that always fails on error and internal interface that always returns the status as 200 with the information whether the operation failed and a message if it did fail. The public interface (whether it's SQL or Python) would use the internal interface and they expose errors in its native way.

Reply:
I like that idea, but anything the Python interface can access would have to be public, ie we would then need to expose the underlying layer.

I'm currently leaning towards figuring out a way to throw errors in SQL objects in a consistent way that the Py Snowpark package can translate with a single utility function. Anyways, like you said, this requires us looking at the SPCS/SQL/Py layers together, so not a blocker for this PR